eclass/dub.eclass

```diff
@@ -0,0 +1,269 @@
+# into DUB_DEPENDENCIES_URIS
+_handle_dub_variables() {
+	[[ -v DUB_DEPENDENCIES && ${DUB_DEPENDENCIES@a} != *a* ]] \
+		&& die "DUB_DEPENDENCIES must be an array"
+
+	[[ ${DUB_OPTIONAL} ]] && return
+
+	BDEPEND+=" ${DUB_DEPEND}"
+
+	[[ ${DUB_DEPENDENCIES[@]} ]] \
+		 && export DUB_DEPENDENCIES_URIS="$(dub_dependencies_uris "${DUB_DEPENDENCIES[@]}")"
+}
+_handle_dub_variables
+unset -f _handle_dub_variables
+
+# @FUNCTION: edub
+# @USAGE: [dub_args]...
+# @DESCRIPTION:
+# A dub wrapper that will automatically set relevant variables and die
+# on failure.
+#
+# Meaningful variables for this function are:
+#
+# $DC - typically set by dlang-r1.eclass or dlang-single.eclass to point
+# to the D compiler and respected by dub.
+#
+# $DFLAGS - dub will use these flags when compiling + linking.  If the
+# variable is unset (empty does not count) this eclass will default to
+# "${DCFLAGS} ${DLANG_LDFLAGS} /-Wno-error/" for its value.
+#
+# $NO_COLOR - if non-empty dub will not print colors
+edub() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	local -x DFLAGS=${DFLAGS-${DCFLAGS} ${DLANG_LDFLAGS} $(dlang_get_wno_error_flag)}
+	local -x DC=${DC} NO_COLOR=${NO_COLOR}
+	# FIXME dub directly calls pkg-config without a way to configure it
+
+	debug-print "${FUNCNAME}: DC=${DC}"
+	debug-print "${FUNCNAME}: DFLAGS=${DFLAGS}"
+
+	set -- dub --verbose "${@}"
+	echo "${@}"
+	"${@}" || die
+}
+
+# @ECLASS_VARIABLE: DUB_LOCAL_REGISTRY
+# @DESCRIPTION:
+# A directory in which dub package archives can be placed for dub to be
+# able to fetch them in the future.  This is useful only if dub _must_
+# perform a fetch operation, perhaps as part of tests, in all other
+# cases DUB_DEPENDENCIES should be used instead.
+#
+# Example:
+# @CODE
+# src_unpack() {
+#   dub_src_unpack
+#   if use test; then
+#       cp "${DISTDIR}/my_package-1.1.0.zip" "${DUB_LOCAL_REGISTRY}"  || die
+#   fi
+# }
+# src_test() {
+#   dub_src_test
+#
+#   # additional test that performs `dub fetch my_package@1.1.0`
+# }
+# @CODE
+DUB_LOCAL_REGISTRY="${EDUB_HOME}/local_registry"
+
+# @FUNCTION: dub_gen_settings
+# @DESCRIPTION:
+# Generate a settings.json file for dub.
+dub_gen_settings() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	mkdir -p "${EDUB_HOME}" || die "Could not create dub home directory"
+	mkdir -p "${DUB_LOCAL_REGISTRY}" || die "Could not create dub registry directory"
+
+	# dub merges the settings from all configuration files so settings
+	# which greatly change the build need to be set to a safe
+	# value. Unfortunately the 'customCachePaths' and 'registryUrls'
+	# settings are cumulative, i.e. they will append to the previous
+	# value, not overwrite it.
+	cat > "${EDUB_HOME}/settings.json" <<-EOF || die "Failed to create dub settings.json file"
+{
+	"registryUrls": [
+		"file://${DUB_LOCAL_REGISTRY}"
+	],
+	"skipRegistry": "standard",
+
+	"customCachePaths": [],
+	"defaultCompiler": "",
+	"defaultArchitecture": ""
+}
```

Comment:
Sam meant that you can indent put a tab in front of thoes lines if you `<<-EOF` instead of `<<EOF`